### PR TITLE
Merge 4.1 to 5.0

### PR DESCRIPTION
**Pull request by Naktibalda**

### CODE REVIEW

**Comment by Naktibalda** [comment]
@TavoNiievez Please check if I haven't missed any types.

File: src/Codeception/Subscriber/FailFast.php [hunk 1]
```diff
@@ -16,12 +17,33 @@ class FailFast implements EventSubscriberInterface
...
+        Events::TEST_ERROR => 'stopOnFail',
+        Events::SUITE_BEFORE => 'cacheSuite'
     ];
 
-    public function stopOnFail(SuiteEvent $event): void
+    private $failureCount = 0;
+
+    private $stopFailureCount;
+
+    private $suiteCache;
```
Comment (TavoNiievez) [inline src/Codeception/Subscriber/FailFast.php:29]: `private ?TestResult $suiteCache = null;`

File: src/Codeception/Subscriber/FailFast.php [hunk 2]
```diff
@@ -16,12 +17,33 @@ class FailFast implements EventSubscriberInterface
...
      */
     protected static array $events = [
-        Events::SUITE_BEFORE => 'stopOnFail',
+        Events::TEST_FAIL => 'stopOnFail',
+        Events::TEST_ERROR => 'stopOnFail',
+        Events::SUITE_BEFORE => 'cacheSuite'
     ];
 
-    public function stopOnFail(SuiteEvent $event): void
+    private $failureCount = 0;
```
Comment (TavoNiievez) [inline src/Codeception/Subscriber/FailFast.php:25]: `int`

File: src/Codeception/Subscriber/FailFast.php [hunk 3]
```diff
@@ -16,12 +17,33 @@ class FailFast implements EventSubscriberInterface
...
+
+    private $stopFailureCount;
```
Comment (TavoNiievez) [inline src/Codeception/Subscriber/FailFast.php:27]: `int`

File: src/Codeception/Subscriber/FailFast.php [hunk 4]
```diff
@@ -16,12 +17,33 @@ class FailFast implements EventSubscriberInterface
...
+
+    private $suiteCache;
+
+    public function __construct($stopFailureCount)
+    {
+        $this->stopFailureCount = (int) $stopFailureCount;
+    }
```
Comment (TavoNiievez) [inline src/Codeception/Subscriber/FailFast.php:34]: how about typing `$stopFailureCount` parameter and adding the `int` cast in src/Codeception/Codecept.php:L132 ?

**Review by TavoNiievez — CHANGES_REQUESTED** [review]

File: src/Codeception/Subscriber/FailFast.php (hunk 4 above)
Comment (Naktibalda) [reply src/Codeception/Subscriber/FailFast.php]: Thanks, I implemented all changes.

**Pull request merged**